### PR TITLE
server <-> browser interoperability test 

### PR DESCRIPTION
**Pull request by kerzhner**
This PR tackles all but the last bullet point of https://github.com/statechannels/statechannels/issues/3383.

The goal of this PR is to create a playground that allows for experimentation on how to share code between server and browser wallets. In this PR:
- `interop-tests` is a new package. The package contains a passing test that established a directly-funded channel between a server wallet and a browser wallet.
- `browser-wallet.ts` in the `xstate-wallet` package is a tiny wallet that re-uses the existing xstate store and chain watcher.
- `browser-wallet.ts` does **not** attempt to describe how reusable protocols should be written. This is to be done in follow up work.

This PR should be follow up with https://github.com/statechannels/statechannels/issues/3400 and https://github.com/statechannels/statechannels/issues/3401.

Fixes https://github.com/statechannels/statechannels/issues/3383 since other issues cover remaining work of https://github.com/statechannels/statechannels/issues/3383.

### CODE REVIEW

**Comment by andrewgordstewart** [comment]
> This PR tackles all but the last bullet point of #3383.

I attempted to simplify the communication here by clarifying what #3383 now covers!

Screen Shot 2021-03-25 at 9 50 41 AM

Now, this PR simply fixes #3383.

File: packages/xstate-wallet/src/browser-wallet.ts [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+      objectives: [],
+      signedStates: []
+    };
+    // Store any new objectives
+    const payloadObjective = payload.objectives?.[0];
+    if (!payloadObjective) {
+      logger.info('No incoming objectives');
+    } else {
+      await this.store.addObjective(payloadObjective);
+    }
```
Comment (andrewgordstewart) [inline packages/xstate-wallet/src/browser-wallet.ts:64]: Is this meant to be permanent? It only stores _one_ new objective, not _all_ new objectives.

File: packages/xstate-wallet/src/browser-wallet.ts [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+    // Store any new states
+    const payloadState = payload.signedStates?.[0];
```
Comment (andrewgordstewart) [inline packages/xstate-wallet/src/browser-wallet.ts:67]: Same as above.

File: packages/xstate-wallet/src/browser-wallet.ts [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+      );
+    }
+    if (response.signedStates[0]) {
+      await this.store.addState(response.signedStates[0]);
+      this.onNewMessage(response);
+    }
+    return response;
+  }
+
+  crankOpenChannelObjective(
```
Comment (andrewgordstewart) [inline packages/xstate-wallet/src/browser-wallet.ts:120]: If this is meant to be a pure function, is there a reason to have it be a (public) method on a class, besides convenience?

File: packages/xstate-wallet/src/browser-wallet.ts [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+      return response;
+    }
+
+    // Should we sign the postfund state?
+    if (
+      BN.gte(channelFunding.amountOnChain, depositInfo.fundedAt) &&
+      latestState.turnNum === 3 &&
+      channel.latestSignedByMe.turnNum === 0
+    ) {
+      const newState = channel.signAndAdd(latestState, pk);
```
Comment (andrewgordstewart) [inline packages/xstate-wallet/src/browser-wallet.ts:154]: How correct is this supposed to be? (It doesn't check that the postFS and the preFS have the same appData & outcome.)

**Review by andrewgordstewart — APPROVED** [review]
Generally LGTM!

Some of the configuration files seem like it contains a lot of "special case rules" for a new package. In the interest of consistency, it would be nice for the config files for a new package to be minimal.

When I think ahead about the directions that this work could take, it feels like filling out `BrowserWallet` _is_ a YABW project. I can imagine a set of refactors to the `server-wallet` that massages business logic out of `Store`, such that the `Store` interface is very simple, and close to the `DexieBackend`. Have you thought about that approach?

File: packages/interop-tests/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,24 @@
+{
+  "extends": "../../tsconfig.json",
+  "compilerOptions": {
```
[the remark this answers is not in the document]
Comment (kerzhner) [reply packages/interop-tests/tsconfig.json:3]: Partially addressed in db284c17b. Also filed https://github.com/statechannels/statechannels/issues/3416

File: packages/interop-tests/.eslintrc.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
+module.exports = {
```
[the remark this answers is not in the document]
Comment (kerzhner) [reply packages/interop-tests/.eslintrc.js:1]: Partially addressed in db284c17b. Also filed https://github.com/statechannels/statechannels/issues/3416

File: packages/interop-tests/tsconfig.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (kerzhner) [reply packages/interop-tests/tsconfig.json:3]: Yes, these are different. Some of the unused options are cleaned up in 23f9eb31f.

File: packages/xstate-wallet/src/browser-wallet.ts (hunk 1 above)
Comment (kerzhner) [reply packages/xstate-wallet/src/browser-wallet.ts:64]: This is not meant to be permanent. This implementation is the quick and dirty implementation to get the interop tests to pass and unblock https://github.com/statechannels/statechannels/issues/3401 and https://github.com/statechannels/statechannels/issues/3400.

File: packages/xstate-wallet/src/browser-wallet.ts (hunk 2 above)
Comment (kerzhner) [reply packages/xstate-wallet/src/browser-wallet.ts:67]: https://github.com/statechannels/statechannels/pull/3402#discussion_r603335387

File: packages/xstate-wallet/src/browser-wallet.ts (hunk 3 above)
Comment (kerzhner) [reply packages/xstate-wallet/src/browser-wallet.ts:120]: https://github.com/statechannels/statechannels/pull/3402#discussion_r603335387

File: packages/xstate-wallet/src/browser-wallet.ts (hunk 4 above)
Comment (kerzhner) [reply packages/xstate-wallet/src/browser-wallet.ts:154]: https://github.com/statechannels/statechannels/pull/3402#discussion_r603335387

File: packages/xstate-wallet/src/chain.ts [hunk 7]
```diff
@@ -325,8 +325,13 @@ export class ChainWatcher implements Chain {
...
   public get selectedAddress(): string | null {
-    if (this.mySelectedAddress === null && window && window.ethereum) {
-      this.mySelectedAddress = window.ethereum.selectedAddress ?? null;
+    if (this.mySelectedAddress === null) {
+      if (window && window.ethereum) {
+        this.mySelectedAddress = window.ethereum.selectedAddress ?? null;
+      } else {
+        // TODO: remove this ganache case hardcoding
+        this.mySelectedAddress = '0xd4Fa489Eacc52BA59438993f37Be9fcC20090E39';
+      }
```
[the remark this answers is not in the document]
Comment (kerzhner) [reply packages/xstate-wallet/src/chain.ts]: Thanks for pointing this out. Addressed in ce399d163.

**Comment by kerzhner** [comment]
> When I think ahead about the directions that this work could take, it feels like filling out `BrowserWallet` _is_ a YABW project. I can imagine a set of refactors to the `server-wallet` that massages business logic out of `Store`, such that the `Store` interface is very simple, and close to the `DexieBackend`. Have you thought about that approach?

This is a great question. It is not clear to me what the browser wallet will look like yet. My goal with this PR and [the follow up PR](https://github.com/statechannels/statechannels/pull/3410) is to create an sandbox for the team to experiment with approaches to writing a the browser wallet.

**Pull request merged**